Run normal pytest tests in pyodide

joemarshall opened the pull request:
This is an extra command line parameter which tells the pytest plugin that you want to run a normal pytest set of tests (i.e. non-pyodide aware, just from any package) but run each test in pyodide.

This means that if you have the pytest plugin installed and working, you can run:

```
# first make a wheel from the current project in dist subfolder, this auto-copies that
pyodide build 

# this autocopies <test_folder> to pyodide and then runs each collected test in pyodide. 
pytest --run-in-pyodide --runtime=chrome test_folder
```
and you should see almost exactly the output you would see if you ran 
`pytest test_folder` normally, but with the tests themselves running in pyodide on a browser.

hoodmane commented:
Thanks @joemarshall this sounds great. How do you deal with dependencies? Maybe we could pass a `requirements.txt` file and a wheels directory where the wheels directory has to include a wheel for each requirement missing from the Pyodide repodata.json file?

hoodmane commented:
Or maybe we could make a virtual environment and ship the site_packages directory of the environment into the browser, and install the dependencies with pip. That's more similar to my command line runner route.

joemarshall commented:
Right now it just installs any wheels in the dist directory. So if you have deps that micropip won't install automatically, you can 
put them in that folder as wheels.

Thinking about it though, I'm not sure how dependency ordering will work - if micropip installs dependencies when you install a wheel, then it in theory should just be a matter of making the wheel require the correct deps. But, a question - if you give micropip a list of wheels, does it correctly resolve them no matter what order they're in? i.e. if I give it main.whl and dep1.whl
in a single install command, will it download just main.whl first and then fail to find dep1, or does it download all given wheels and then try resolving dependencies?

hoodmane commented:
> But, a question - if you give micropip a list of wheels, does it correctly resolve them no matter what order they're in?

I'm not sure...

joemarshall commented:
I just had a think about this - I think the right thing to do here is do a bit of work on pyodide build to enable it to build a project and collect (and build if needed) all deps into the dist folder. That way once you have the project working with pyodide build, you both have all required dependencies for running it in pyodide or jupyterlite easy to hand, and also for running pytest on it. Needs a little thought as to how to specify what doesn't need building (e.g. if building things which support jupyter, everything is a hell of a lot easier if you tell it to ignore various jupyter deps which themselves have complex dependency trees, which are implemented by jupyterlite anyway) 


I've got to make roughly that functionality work anyway for what I'm working on, and the client on this is keen for things to upstream where possible as opposed to having to maintain complex build scripts, so it could be a win win.

joemarshall commented:
Oh, I realized why this wouldn't pass tests - I had a bunch of broken testing stuff left in github actions. Sorry, fixed now.

@hoodmane I think subject to the dependency resolution stuff going into pyodide build, this is hopefully ready to go

joemarshall commented:
@hoodmane Did you get a chance to look at this yet?  I can already see places where this might be useful (e.g. https://github.com/pyodide/pyodide/pull/2967 ) and I could do with it in for stuff I'm up to.

hoodmane commented:
Started looking at it should get a review out soon. I tend to forget about prs on these other repos...

rth commented on pytest_pyodide/hook.py:
```diff
@@ -172,19 +200,55 @@ def _get_item_position(item):
+            # pass this test to pyodide runner
+            # first: make sure that pyodide has the current folder copied over
+            itemPath = Path(item.path)
+            copy_files = list(itemPath.parent.glob("**/*"))
```
What if there are subdirectories? This should probably recursively list the files no?

joemarshall replied:
It does, it's a pathlib.path, glob is recursive by default

rth reviewed (commented):
Thanks for working on it @joemarshall it's indeed a very useful feature!
A few comments about the code below.

About the CLI API. So there are indeed packages that have a separate test folder. But many also include the tests folders directly inside the module (also docstests). Would this work in that case?

ryanking13 commented:
Thanks for the work @joemarshall!

If possible, I think it would be nice to split this PR into two PRs:

1) Copying local files into the Pyodide virtual file system [^1]
2) Adding `--run-in-pyodide` CLI flag.

So we can discuss seperately about the requirements for each PR. If it is too hard to split, it is okay to say no :)

On the other hand, we probably need to discuss 1) when should people use  pyodide cmdline runner and 1) when should they use `pytest-pyodide` to run in-browser tests.

[^1]: We can probably also check what @DerThorsten did in [pyjs-code-runner](https://github.com/emscripten-forge/pyjs-code-runner) to copy local files into the file system.

joemarshall commented on pytest_pyodide/pytest_in_pyodide.py:
```diff
@@ -0,0 +1,174 @@
+    )
+    if len(ret_error) != 0:
+        print("ERR:", ret_error, "\n*******************")
+        ret_error_lines = ret_error.splitlines()
```
Yes, great idea, done it now.

joemarshall commented:
@ryanking13 Actually splitting it would be a bit fiddly, but I have split the key logic in the code into two files - `run_tests_inside_pyodide.py` which calls tests inside pyodide, and `copy_files_to_pyodide.py` which does the copying stuff. I also added a test decorator for copying files, which allows you to use the copy files functionality in other tests.

I didn't yet add a command line for copying files, but I think that would be a pretty trivial addition, I just ran out of time. If you're okay, I can merge this version now it tests okay, and add the copying files command line later?

ryanking13 commented on tests/test_copy_files.py:
```diff
@@ -0,0 +1,147 @@
+        pytest.fail(
+            "Copy files to pyodide should fail if source is above current directory"
+        )
+    except RuntimeError:
```
Maybe use `pytest.raises` instead?

ryanking13 commented on tests/test_copy_files.py:
```diff
@@ -0,0 +1,147 @@
+        old_dir = Path.cwd()
+        try:
+            os.chdir(td)
+            run([sys.executable, "-m", "pip", "download", "pyodide_http"])
```
Maybe let's embed test wheel file inside the repository instead of downloading it every time.

ryanking13 reviewed (approved):
Thanks, @joemarshall. I have several comments, but it generally looks like a nice improvement.

>  add the copying files command line later?

Yes, of course.

joemarshall commented on pytest_pyodide/hook.py:
```diff
@@ -41,6 +47,11 @@ def _filter_runtimes(runtime: list[str]) -> tuple[bool, set[str]]:
     return run_host, runtime_filtered
 
 
+def pytest_unconfigure(config):
```
it appears to be called on keyboardinterrupt. You could probably kill it somehow and avoid it, but as long as python quits nicely it should be.
